pkg/plugins/diff/delta_diff_server/src/main.rs

```diff
@@ -0,0 +1,211 @@
+        }
+        let mut right_table = right_table_res?;
+
+        // If the table exists on the right, but doesn't exist on the left -> table created.
+        if left_table_res.is_err() {
+            let hist = delta_ops::history(&mut right_table, None).await?;
+            let ans = reverse_and_create_table_ops(hist, right_table.version());
+            return Ok(Response::new(DiffResponse { entries: ans, diff_type: differ::DiffType::Created as i32 }))
+        }
+        let mut left_table= left_table_res?;
+
+        // todo: if base table exists, use it's version to start the comparison instead of going from most recent to oldest.
+        // let mut base_table: Option<DeltaTable> = None;
+        // if let Some(table_path) = base_table_path {
+        //     base_table = Option::from(delta_ops::get_delta_table(&s3_config_map, &diff_props.repo, table_path).await?);
+        // }
+
+        let left_table_history = delta_ops::history(&mut left_table, None).await?;
+        let right_table_history = delta_ops::history(&mut right_table, None).await?;
+
+        let mut left_history_version = HistoryAndVersion{ history: left_table_history, version: left_table.version() };
+        let mut right_history_version = HistoryAndVersion{ history: right_table_history, version: right_table.version() };
+
+        let ans = compare(&mut left_history_version,  &mut right_history_version).unwrap();
+        return Ok(Response::new(DiffResponse { entries: ans, diff_type: differ::DiffType::Changed as i32 }))
+    }
+
+    async fn show_history(&self, _request: Request<HistoryRequest>) -> Result<Response<HistoryResponse>, Status> {
+        todo!()
+    }
+}
+
+struct HistoryAndVersion {
+    pub history: Vec<Map<String, Value>>,
+    pub version: DeltaDataTypeVersion,
+}
+
+impl HistoryAndVersion {
+    fn get_earliest_version(&self) -> i64 {
+        self.version + 1 - i64::try_from(self.history.len()).unwrap()
+    }
+
+    fn get_lower_limit(&self, other: &HistoryAndVersion) -> i64 {
+        let self_early = self.get_earliest_version();
+        let other_early = self.get_earliest_version();
+        if self_early > other.version { self_early } else { other_early }
+    }
+}
+
+fn reverse_and_create_table_ops(mut hist: Vec<Map<String, Value>>, table_version: DeltaDataTypeVersion) -> Vec<TableOperation> {
+    hist.reverse();
+    let ops = OpTypes::new();
+    let mut ans: Vec<TableOperation> = Vec::with_capacity(hist.len());
+    for m in hist {
+        let table_op = utils::construct_table_op(&m, table_version, &ops);
+        ans.push(table_op);
+    }
+    return ans;
+}
+
+fn compare(left_history_version: &mut HistoryAndVersion,
+           right_history_version: &mut HistoryAndVersion,) -> Result<Vec<TableOperation>, Error> {
+    // The lower limit of the two (earliest common version):
+    let lower_limit = right_history_version.get_lower_limit(&left_history_version);
```

Comment:
I must be missing something, so let me know where I'm wrong...
Base has version 2. Right created a branch and made 2 changes to the table -> version 4. Left worked on the base branch and made a single change to the table -> version 3. Starting to compare from version 3 is wrong. How can we not use the base here?

Reply:
this is not the case. the naming is confusing, I'll change it.
the version that will be returned is the earliest version that is found on both the right and left.